### ISSUE TITLE
Package cannot export `package.json` after version 0.3.0

### ISSUE THREAD

**Issue by xfsnowind** [web-infra-dev/rspack#4541]
### System Info

System:
    OS: macOS 14.0
    CPU: (10) arm64 Apple M1 Pro
    Memory: 66.09 MB / 16.00 GB
    Shell: 5.9 - /bin/zsh
  Binaries:
    Node: 21.1.0 - /opt/homebrew/bin/node
    Yarn: 1.22.19 - /opt/homebrew/bin/yarn
    npm: 10.2.0 - /opt/homebrew/bin/npm
    pnpm: 8.7.0 - ~/Library/pnpm/pnpm
  Browsers:
    Brave Browser: 115.1.56.11
    Chrome: 119.0.6045.105
    Edge: 114.0.1823.55
    Safari: 17.0
  npmPackages:
    @rspack/cli: 0.3.4 => 0.3.4
    @rspack/core: 0.3.4 => 0.3.4

### Details

Currently we are using [cypress-rspack-dev-server](https://github.com/th3fallen/cypress-rspack-dev-server) as the dev-server for cypress, when we upgraded rspack from version 0.2 to 0.3.0, it reports the error:
```
Error [ERR_PACKAGE_PATH_NOT_EXPORTED]: Package subpath './package.json' is not defined by "exports" in /Users/rodrigofaria/Documents/frontend/fleetweb-wt/copy2/node_modules/@rspack/core/package.json
```

And I think it's caused by [restricted accessing to internal modules](https://www.rspack.dev/blog/announcing-0.3.html#access-to-internal-modules-restricted), which explicitly defines the `exports` in related packages' `package.json` files.

It causes the `package.json` file could not be resolved by Nodejs's `require.resolve`:
<img width="591" alt="image" src="https://github.com/web-infra-dev/rspack/assets/2863328/a29dabdb-8bc9-43e1-a920-102ce832bb76">

Source code is [here](https://github.com/th3fallen/cypress-rspack-dev-server/blob/08433e4c925c773db5688cf60bf87d089e6cfaae/src/helpers/sourceRelativeRspackModules.ts#L120), which copy the same logic from [webpack-dev-server](https://github.com/cypress-io/cypress/blob/5dbebe6e5e7fca3af9716bfa8b80830b87641cf4/npm/webpack-dev-server/src/helpers/sourceRelativeWebpackModules.ts#L127).

I manually add one extra export for this `package.json` and can fix this problem, but not sure it's the best way to do this:
![image](https://github.com/web-infra-dev/rspack/assets/2863328/abd55419-8e83-4445-a742-a5a12e266da8)

The error is reported by my colleague already: [here](https://github.com/th3fallen/cypress-rspack-dev-server/issues/8)

### Reproduce Steps

1. install the lib cypress, cypress-rspack-dev-server
2. install the rspack core and cli with version 0.3.0
3. set the cypress config file:
```
import { devServer } from 'cypress-rspack-dev-server'
import { defineConfig } from 'cypress'

export default defineConfig({
  video: false,
  component: {
    viewportWidth: 1280,
    viewportHeight: 720,
    experimentalSingleTabRunMode: true, // improves performance on cypress run (good for CI)
    experimentalMemoryManagement: true, // should solve "Out of Memory" issues
    devServer(devServerConfig) {
      return devServer({
        ...devServerConfig,
        framework: 'react',
        rspackConfig: require('./webpack.config.js'),
      })
    },
  },
})
```
4. run the command `pnpm cypress run --component --config-file ./my-config.ts`

**Comment by hardfist** [comment]
makes sense, it seems we should export package.json

**Comment by xfsnowind** [comment]
> make sense, it seems we should export package.json

Great, could I create a PR for that?

**Comment by hardfist** [comment]
fixed in #4543
